### PR TITLE
fix: keep raw header name

### PR DESCRIPTION
**Pull request by tsctx**
Follow up of #3160.
In the following cases. Lose the original raw header name.
```js
await fetch(new Request(url, { headers: [['Content-Type', 'text/plain']] }), { method: 'GET' })
```

### CODE REVIEW

File: lib/web/fetch/request.js [hunk 1]
```diff
@@ -477,9 +477,11 @@ class Request {
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers instanceof HeadersList) {
-        for (const { 0: key, 1: val } of headers) {
-          // Note: The header names are already in lowercase.
-          headersList.append(key, val, true)
+        const rawList = headersList.entriesRawList
```
Comment (tsctx) [inline lib/web/fetch/request.js:480]:
```suggestion
        const rawList = headers.entriesRawList
```

File: test/fetch/headers-case.js [hunk 2]
```diff
@@ -1,30 +1,32 @@
...
     { 'Content-Type': 'text/plain' },
     [['Content-Type', 'text/plain']]
   ]) {
-    await fetch(`http://localhost:${server.address().port}`, {
-      headers
-    })
+    await fetch(url, { headers })
   }
+  // see https://github.com/nodejs/undici/pull/3183
+  await fetch(new Request(url, { headers: new Headers([['Content-Type', 'text/plain']]) }), { headers: undefined })
```
Comment (tsctx) [inline test/fetch/headers-case.js:31]:
```suggestion
  await fetch(new Request(url, { headers: new Headers([['Content-Type', 'text/plain']]) }), { method: 'GET' })
```

File: lib/web/fetch/request.js [hunk 3]
```diff
@@ -477,9 +477,11 @@ class Request {
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers instanceof HeadersList) {
-        for (const { 0: key, 1: val } of headers) {
-          // Note: The header names are already in lowercase.
-          headersList.append(key, val, true)
+        const rawList = headers.entriesRawList
+        // TODO: optimizing this.
```
Comment (tsctx) [inline lib/web/fetch/request.js:481]:
```suggestion
```

File: test/fetch/headers-case.js [hunk 4]
```diff
@@ -1,30 +1,32 @@
...
     { 'Content-Type': 'text/plain' },
     [['Content-Type', 'text/plain']]
   ]) {
-    await fetch(`http://localhost:${server.address().port}`, {
-      headers
-    })
+    await fetch(url, { headers })
   }
+  // see https://github.com/nodejs/undici/pull/3183
+  await fetch(new Request(url, { headers: new Headers([['Content-Type', 'text/plain']]) }), { method: 'GET' })
```
Comment (tsctx) [inline test/fetch/headers-case.js:31]:
```suggestion
  await fetch(new Request(url, { headers: [['Content-Type', 'text/plain']] }), { method: 'GET' })
```

File: lib/web/fetch/headers.js [hunk 5]
```diff
@@ -259,6 +259,18 @@ class HeadersList {
     return headers
   }
 
+  get entriesRawList () {
```
Comment (KhafraDev) [inline lib/web/fetch/headers.js:262]: this could just be an iterator

**Review by KhafraDev — APPROVED** [review]

**Review by metcoder95 — APPROVED** [review]

File: lib/web/fetch/headers.js (hunk 5 above)
Comment (tsctx) [reply lib/web/fetch/headers.js:262]: thank you for your great idea.

File: lib/web/fetch/headers.js [hunk 6]
```diff
@@ -259,6 +259,18 @@ class HeadersList {
...
+    const headers = []
+
+    if (this[kHeadersMap].size !== 0) {
+      for (const { name, value } of this[kHeadersMap].values()) {
+        headers.push([name, value])
+      }
+    }
+
+    return headers
+  }
```
Comment (tsctx) [inline lib/web/fetch/headers.js:272]:
```suggestion
  rawValues () {
    return this[kHeadersMap].values()
  }
```

File: lib/web/fetch/request.js [hunk 7]
```diff
@@ -477,9 +477,10 @@ class Request {
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers instanceof HeadersList) {
-        for (const { 0: key, 1: val } of headers) {
-          // Note: The header names are already in lowercase.
-          headersList.append(key, val, true)
+        const rawList = headers.entriesRawList
+        for (let i = 0; i < rawList.length; ++i) {
+          const { 0: key, 1: value } = rawList[i]
+          headersList.append(key, value, false)
```
Comment (tsctx) [inline lib/web/fetch/request.js:483]:
```suggestion
        for (const { key, value } of headers.rawValues()) {
          headersList.append(key, value, false)
```

File: lib/web/fetch/request.js [hunk 8]
```diff
@@ -477,9 +477,8 @@ class Request {
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers instanceof HeadersList) {
-        for (const { 0: key, 1: val } of headers) {
-          // Note: The header names are already in lowercase.
-          headersList.append(key, val, true)
+        for (const { key, value } of headers.rawValues()) {
+          headersList.append(key, value, false)
```
Comment (tsctx) [inline lib/web/fetch/request.js:481]:
```suggestion
        for (const { name, value } of headers.rawValues()) {
          headersList.append(name, value, false)
```

File: lib/web/fetch/headers.js [hunk 9]
```diff
@@ -259,6 +259,10 @@ class HeadersList {
     return headers
   }
 
+  rawValues () {
```
Comment (mcollina) [inline lib/web/fetch/headers.js:262]: is this part of the spec?

Comment (tsctx) [reply lib/web/fetch/headers.js:262]: No, non-spec.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**